src/Sylius/Bundle/ShopBundle/EventListener/OrderPromotionIntegrityChecker.php

```diff
@@ -61,24 +51,21 @@ public function check(ResourceControllerEvent $event): void
 
         Assert::isInstanceOf($order, OrderInterface::class);
 
-        // we create a new promotion collection and remove them from cart
-        // so we can verify with original conditions (without the price being applied before check)
-
-        $promotions = $order->getPromotions()->toArray();
+        $previouslyAppliedPromotionsCodeToNameMap = $this->createPromotionCodeToNameMap($order);
 
-        if ($this->promotionApplicator !== null) {
-            foreach ($promotions as $promotion) {
-                $this->promotionApplicator->revert($order, $promotion);
-                $order->removePromotion($promotion);
-            }
+        if ($this->orderProcessor !== null) {
```

Comment:
It is not optional in the constructor and without it the whole thing won't work.